README.md
# ROXconnector API server

A fast and flexible Node.js server for delivering RESTful APIs based on JSON data serialization - with plugin support.
The endpoint definition resides inside the servers config file - not within the code.
Endpoints can be configured to either invoke a process, be relayed to another server via http or mapped onto a plugin function.

## Open Source

Copyright (C) 2019  droxIT GmbH - devs@droxit.de

ROXconnector is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.

You have received a copy of the GNU General Public License
along with this program ([plain text](COPYING), [Markdown](COPYING.md)). See also <http://www.gnu.org/licenses/>.

## Why use ROXconnector?

There are serveral reasons for using ROXconnector as your API gateway:

* Easy setup: ROXconnector is operational within minutes
* Integration with other services: you can use ROXconnector to map REST calls to other services via HTTP
* Extensibility: add new functionality to your endpoints with your own Node modules - access your own code from within the API server itself
* REST setup inside the configuration: define your whole API in a single JSON file. Simply restart ROXconnector to see the changes
* In essence: be fast and flexible

## Installation

The following software components need to installed before proceeding:

* [Node.js 8.10.0](https://nodejs.org) or higher - a JavaScript runtime.
* [npm](https://www.npmjs.com/) - a package manager for JavaScript applications.

Set up an application folder for ROXconnector to reside in - we will use `/opt/roxconnector` as an example.
Then you simply have to extract the package and install the dependencies via `npm`.

```bash
mkdir /opt/roxconnector/
cd /opt/roxconnector
tar xf /tmp/roxconnector-0.4.3.tgz
npm install
```

This will install the node dependencies under `node_modules` within the roxconnector folder. The dependencies can also be installed system wide by providing the `-g` flag.

```bash
npm install -g
```

This step usually requires superuser priviledges.

## Usage

You will need a working configuration file to run `ROXconnector` - explained below.

```bash
cd /opt/roxconnector
node startServer.js path/to/config.js
```

## Configuration

ROXconnector is configured via one central JSON config file which contains all needed parameters.
The file is divided into three segments: SYSTEM, REST and PLUGINS.

### SYSTEM

```json
{
  "SYSTEM": {
    "port": 7475,
    "upload_limit": "1000kb",
    "loglevel": "info",
    "logfile": "../logs/roxconnector.log"
  }
}
```

The config's **SYSTEM** portion contains startup information for the server process. The mandatory _port_ parameter defines the TCP port to listen on, whereas _upload_limit_ adjusts the maximum upload size.
_loglevel_ and _logfile_ parameters are optional. The default log level is _warn_ and output is directed to _stderr_ if the _logfile_ option is missing. The server uses [Bunyan](https://github.com/trentm/node-bunyan)
as its logging facility. Bunyan's output is well suited for automated handling but not as well readable by humans. Therefore the bunyan package provides a binary that converts the logs into a format that's more legible.
If you have installed the node dependencies globally the bunyan application should be inside your path environment. Otherwise you can call it from the `node_modules` folder:

```bash
# pipe console output into bunyan
node start_server.js | node_modules/bunyan/bin/bunyan
# or view logs on disk
cat /var/log/roxconnector/server.log | node_modules/bunyan/bin/bunyan
```

### REST

```json
"REST": {
  "GET": {
    "mybackend": {
      "handler": {
        "type": "http",
        "options": {
          "host": "my.backend.de",
          "path": "/"
        }
      }
    }
  },
  "POST": {
    "echo": {
      "handler": {
        "type": "process",
        "command": ["cat", "-"]
      }
    },
    "nested": {
      "path": {
        "handler": {
          "type": "plugin",
          "name": "myplugin",
          "function": "some_function"
        }
      }
    }
  }
}
```

The **REST** property is root of all RESTful service endpoints.
Endpoints can be defined freely in the method subtrees ('GET', 'POST' etc.). In the example POSTing to `localhost:8160/echo` would return the data we send with the POST request.
Endpoint handlers can either be of type process or http. Type process means that every time a request is sent to this endpoint the command array is executed as a subprocess and
anything printed to stdout during the subprocess' life time will be sent back to the client once the process terminates. Data sent via POST or PUT will be fed into the handler
process' stdin. The command structure looks like this:
 
    [ "executable", "arg1", "arg2" ... ]

Http handlers try to communicate with another service via http. The options field is fed to Node.js' native [http.request method](https://nodejs.org/dist/latest-v6.x/docs/api/http.html#http_http_request_options_callback)
therefore all options the method supports can be used. Besides type and options there is another optional field called data. This determins how data sent with the client request are handled. Without the field no data
is sent to the http service the handler points to. Putting a '-' as the field's value will cause data sent by the client to be relayed. Any other value in data will be sent with every request to the http service while
any data the client might send will be discarded.


Plugin handlers are straight forward. In this example a `POST` call to the path `/nested/path/` would result in a call to `some_function` from the plugin `myplugin` (more on plugins later).

### PLUGINS

```json
"PLUGINS": {
  "myplugin": {
    "path": "./plugins/myplugin.js",
    "params": {
      "plugin_parameter": "some_information"
    }
  }
}
```
Any plugin that is supposed to be loaded needs to be present in this section with their path and optionally a params dictionary that is passed on to the plugin upon loading.

## Writing plugins

An ROXconnector plugin is a Node.js module that adheres to certain export conventions and exports an `init` function. `module.exports` must be a function that takes a single argument
and registers its endpoint handlers within this argument:

```javascript
module.exports = function (container) {
    container.init = init;
    container.some_function = sfunc;
    // ....
}
```
The `init` function will be called with the parameters from the config when the module is loaded at startup. Our example config would trigger this call:

```javascript
init( { plugin_parameter: "some_information" } );
```

Use this to initialize your plugin.

All exported functions can be mapped to REST endpoints in the ROXconnector config file. In the example config above we mapped `/nested/path` to `some_function`.
Assuming the following client request:

```bash
curl -XPOST localhost:7475/nested/path -H 'Content-Type: application/json' -d '{ "arg": 213 }'
```

the resulting function call would be `sfunc({arg: 123})`.


## droxit-api.service and install\_service.sh
The roxconnector.service file allows the api server to be deployed as a [systemd](https://en.wikipedia.org/wiki/Systemd) service. To depoy the service run

    sudo install_service.sh /path/to/config.js

from within this directory. The configuration file needs to point to a valid config. The file's location is written into the service and therefore is needed at this location for the server
to start up.



CHANGELOG.md
# v0.4.3

* No changes

# v0.4.2

## New

* License switched to GPL
* upload limit is configurable under SYSTEM.upload\_limit

# v0.4.1

## New

* Endpoints are now valid with or without trailing slashes.

## Fixed

* Better error handling when initializing a plugin.


test_api_server.js
var http = require('http');
var request = require('supertest');
var apiSrv = require('./droxitApi.js');
var Emitter = require('events').EventEmitter;
var emitter = new Emitter();

var apiPort = 8601;
var dummyPort = 8602;
var dummyMsg = '{"re": "ality"}'; 

function kthxbye() {
	console.log('----- all test successful -----');
	process.exit(0);
}
apiConfig = {
	SYSTEM: {
		port: apiPort
	},
	REST: {
		GET: {
			fortune_cookie: {
				handler: {
					type: 'http',
					options: {
						host: 'localhost',
						port: dummyPort,
						path: '/'
					}
				}
			}
		},
		POST: {
			echo: {
				handler: {
					type: 'process',
					command: ['cat', '-']
				}
			},
			fortune_cookie: {
				handler: {
					type: 'http',
					data: '-',
					options: {
						method: 'POST',
						host: 'localhost',
						port: dummyPort,
						path: '/'
					}
				}
			}
		}
	}
};

// we start a little dummy server to test the http-handler functionality
dummy = new http.Server().on('request', (req, res) => {
	if (req.method === 'GET') {
		res.writeHead(200, {'Content-Length': dummyMsg.length, 'Content-Type': 'application/json'});
		res.write(dummyMsg);
		res.end();
	} else if(req.method === 'POST') {
		var data = '';
		req.on('data', function(chunk) { data += chunk; });
		req.on('end', function() {
			res.writeHead(200, {'Content-Length': data.length, 'Content-Type': 'application/json'});
			res.write(data);
			res.end();
		});
	} else {
		res.writeHead(404);
		res.end();
	}
});
dummy.listen(dummyPort);

srv = apiSrv.new(apiConfig);
request = request(srv);

console.log('----- running tests -----');
console.log('** making a valid GET request **');
request.get('/fortune_cookie')
	.expect('Content-Type', /application\/json/)
	.expect(200, dummyMsg)
	.end(function(err, res) {
		if(err)
			throw err;
		console.log('** got the right answer **');
		emitter.emit('test1');
	});

emitter.on('test1', function() {
	console.log('** GETting invalid path **');
	request.get('/fortune_rookie')
		.expect('Content-Type', /application\/json/)
		.expect(404)
		.end(function(err, res) {
			if(err)
				throw err;
			console.log('** got 404 - good **');
			emitter.emit('test2');
		});
});

emitter.on('test2', function() {
	var testMsg = '{"msg":"blorp"}';
	console.log('** POSTing something to a process **');
	request.post('/echo')
		.send(testMsg)
		.set('Content-Type', 'application/json')
		.expect('Content-Type', /application\/json/)
		.expect(200, testMsg)
		.end(function(err, res) {
			if(err)
				throw err;
			console.log('** got the correct response **');
			emitter.emit('test3');
		});
});

emitter.on('test3', function() {
	var testMsg = '{"msg":"blorp"}';
	console.log('** POSTing something to a http handler **');
	request.post('/fortune_cookie')
		.send(testMsg)
		.set('Content-Type', 'application/json')
		.expect('Content-Type', /application\/json/)
		.expect(200, testMsg)
		.end(function(err, res) {
			if(err)
				throw err;
			console.log('** got the correct response **');
			kthxbye();
		});
});



test/test_api_server.js
/*
 *
 * |------------------- OPEN SOURCE LICENSE DISCLAIMER -------------------|
 * |                                                                      |
 * | Copyright (C) 2019  droxIT GmbH - devs@droxit.de                     |
 * |                                                                      |
 * | This file is part of ROXconnector.                                   |
 * |                                                                      |
 * | ROXconnector is free software: you can redistribute it and/or modify |
 * | it under the terms of the GNU General Public License as published by |
 * | the Free Software Foundation, either version 3 of the License, or    |
 * | (at your option) any later version.                                  |
 * |                                                                      |
 * | This program is distributed in the hope that it will be useful,      |
 * | but WITHOUT ANY WARRANTY; without even the implied warranty of       |
 * | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the         |
 * | GNU General Public License for more details.                         |
 * |                                                                      |
 * | You have received a copy of the GNU General Public License           |
 * | along with this program. See also <http://www.gnu.org/licenses/>.    |
 * |                                                                      |
 * |----------------------------------------------------------------------|
 */

var http = require('http');
var request = require('supertest');
var apiSrv = require('../src/roxconnector.js');
var Emitter = require('events').EventEmitter;
var emitter = new Emitter();

var apiPort = 8601;
var dummyPort = 8602;
var dummyMsg = '{"re": "ality"}';

function kthxbye() {
	console.log('----- all test successful -----');
	process.exit(0);
}
apiConfig = {
	SYSTEM: {
		port: apiPort
	},
	REST: {
		GET: {
			fortune_cookie: {
				handler: {
					type: 'http',
					options: {
						host: 'localhost',
						port: dummyPort,
						path: '/'
					}
				}
			}
		},
		POST: {
			echo: {
				handler: {
					type: 'process',
					command: ['cat', '-']
				}
			},
			fortune_cookie: {
				handler: {
					type: 'http',
					data: '-',
					options: {
						method: 'POST',
						host: 'localhost',
						port: dummyPort,
						path: '/'
					}
				}
			}
		}
	}
};

// we start a little dummy server to test the http-handler functionality
dummy = new http.Server().on('request', (req, res) => {
	if (req.method === 'GET') {
		res.writeHead(200, {
			'Content-Length': dummyMsg.length,
			'Content-Type': 'application/json'
		});
		res.write(dummyMsg);
		res.end();
	} else if (req.method === 'POST') {
		var data = '';
		req.on('data', function(chunk) {
			data += chunk;
		});
		req.on('end', function() {
			res.writeHead(200, {
				'Content-Length': data.length,
				'Content-Type': 'application/json'
			});
			res.write(data);
			res.end();
		});
	} else {
		res.writeHead(404);
		res.end();
	}
});
dummy.listen(dummyPort);

srv = apiSrv.new(apiConfig);
request = request(srv);

console.log('----- running tests -----');
console.log('** making a valid GET request **');
request.get('/fortune_cookie')
	.expect('Content-Type', /application\/json/)
	.expect(200, dummyMsg)
	.end(function(err, res) {
		if (err)
			throw err;
		console.log('** got the right answer **');
		emitter.emit('test1');
	});

emitter.on('test1', function() {
	console.log('** GETting invalid path **');
	request.get('/fortune_rookie')
		.expect('Content-Type', /application\/json/)
		.expect(404)
		.end(function(err, res) {
			if (err)
				throw err;
			console.log('** got 404 - good **');
			emitter.emit('test2');
		});
});

emitter.on('test2', function() {
	var testMsg = '{"msg":"blorp"}';
	console.log('** POSTing something to a process **');
	request.post('/echo')
		.send(testMsg)
		.set('Content-Type', 'application/json')
		.expect('Content-Type', /application\/json/)
		.expect(200, testMsg)
		.end(function(err, res) {
			if (err)
				throw err;
			console.log('** got the correct response **');
			emitter.emit('test3');
		});
});

emitter.on('test3', function() {
	var testMsg = '{"msg":"blorp"}';
	console.log('** POSTing something to a http handler **');
	request.post('/fortune_cookie')
		.send(testMsg)
		.set('Content-Type', 'application/json')
		.expect('Content-Type', /application\/json/)
		.expect(200, testMsg)
		.end(function(err, res) {
			if (err)
				throw err;
			console.log('** got the correct response **');
			emitter.emit('test4');
		});
});

emitter.on('test4', function() {
	console.log('** GETing something from URL with trailing slash **');
	request.get('/fortune_cookie/')
		.expect('Content-Type', /application\/json/)
		.expect(200, dummyMsg)
		.end(function(err, res) {
			if (err)
				throw err;
			console.log('** got the right answer **');
			kthxbye();
		});
});


droxitApi.js
/*
 * Droxit API server package
 *
 * This server exposes a freely configurable REST service for interaction with the core system
 */
var fs = require('fs');
var express = require('express');
var http = require('http');
var bodyParser = require('body-parser');
var url = require('url');
var spawn = require('child_process').spawn;
var bunyan = require('bunyan');

var exp = module.exports;
var _plugins = {};

function setEndpoints(app, config, logger) {
	// handle errors gracefully
	app.use(function(err, req, res, next) {
		if(err) {
			logger.error(err);
			res.status(err.status).send(err.message);
		} else {
			next();
		}
	});
	// main http handling funcionality - available endpoints are defined in the config file
	app.use('/', function(req, res, next) {
		// setting headers to enable CORS (Cross Origin Ressource Sharing)
		res.header("Access-Control-Allow-Origin", req.get('Origin') || '*');
		res.header("Access-Control-Allow-Credentials", "true");
		res.header("Access-Control-Allow-Methods", "GET, HEAD, PUT, PATCH, POST, DELETE, OPTION");
		res.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Accept, Content-Type");

		res.append('Content-Type', 'application/json');
		logger.debug({request: req}, 'received request');

		if(req.method in config.REST) {
			// we only accept JSON encoded data bodies
			if(req.method === 'POST' || req.method === 'PUT') {
				if( !('content-type' in req.headers) ||
					(req.headers['content-type'] !== 'application/json'
						&& req.headers['content-type'] !== 'application/x-www-form-urlencoded')) {
					res.status(415).send('Invalid media type. MIME Type application/json required!');
					return;
				}
			}

			// check the config if the requested path is a valid endpoint
			var endpoint = config.REST[req.method];
			var path = url.parse(req.url).pathname.slice(1).split('/');
			for(var p in path) {
				if(path[p] in endpoint)
					endpoint = endpoint[path[p]];
				else {
					res.status(404).send('{"reason": "Invalid path"}');
					return;
				}
			}
			if('handler' in endpoint) {
				if (endpoint.handler.type == 'process') {
					var jobConfig = endpoint.handler.command;
					var childProcess = spawn(jobConfig[0], jobConfig.slice(1));
					var childPid = childProcess.pid;
					var response = '';
					var error = '';

					if(req.body) {
						childProcess.stdin.write(JSON.stringify(req.body));
						childProcess.stdin.end();
					}

					childProcess.stdout.on('data', (data) => {
						response += data;
					});

					// logging stderr, to know if something bad happened
					childProcess.stderr.on('data', (data) => {
						error += data;
					});

					childProcess.on('exit', (code) => {
						logger.info({pid: childPid, exitCode: code}, 'child process exited');
						if(error) {
							try {
								logger.error({childPid: childPid, err_msg: error, requestBody: req.body, endpoint: '/' + path.join("/")}, 'handler process reported an error');
								err_obj = JSON.parse(error);
								res.status(err_obj.code).send(err_obj.reason);
							} catch(e) {
								res.status(500).send('{"reason": "internal server error"}');
							}
						} else
							res.send(response);
					});
				} else if (endpoint.handler.type == 'http') {
					if ('options' in endpoint.handler) {
						var data = '';
						// send a request to the handler
						var handlerReq = http.request(endpoint.handler.options, (r) => {
							var data = '';
							r.on('data', (chunk) => {
								data += chunk;
								logger.debug('received chunk: %s', chunk);
							});
							r.on('end', () => {
								logger.debug('sending data');
								res.send(data);
							});
						});
						handlerReq.on('error', (e) => {
							logger.error(e);
							res.status(500).send('{"reason": "internal server error"}');
						});
						if ('data' in endpoint.handler) {
							// use user data if any
							if (endpoint.handler.data === '-') {
								data = req.body;
							} else {
								data = endpoint.handler.data;
							}
						}

						if (data) {
							logger.debug({data: data}, "sending data to http handler");
							handlerReq.write(JSON.stringify(data));
						}
						handlerReq.end();
					} else {
						logger.error({path: "/" + path.join("/")}, "faulty configuration: endpoint is missing options field");
						res.status(500).send('{"reason": "internal server error"}');
					}
				} else if(endpoint.handler.type === 'plugin') {
					if(('name' in endpoint.handler) && ('function' in endpoint.handler)) { 
						var pname = endpoint.handler.name;
						var func = endpoint.handler['function'];

						if((pname in _plugins) && (func in _plugins[pname])) {
							var data = {};
							if(req.body) {
								data = req.body;
							}
							_plugins[pname][func](data, function(e, r) {
								if(e) {
									logger.error({plugin: pname, func: func, error: e}, "plugin returned an error");
									res.status(e.code).send(e.message);
								} else {
									res.send(r);
								}
							});
						} else {
							logger.error({path: "/" + path.join("/")}, "faulty configuration: plugin or function missing");
							res.status(500).send('{"reason": "internal server error"}');
						}
					} else {
						//TODO continue here
						logger.error({path: "/" + path.join("/")}, "faulty configuration: plugin handler is missing name or function argument");
						res.status(500).send('{"reason": "internal server error"}');
					}
					var func = endpoint.handler.function;
				} else {
					logger.error({path: "/" + path.join("/")}, "faulty configuration: invalid endpoint handler type");
					res.status(500).send('{"reason": "internal server error"}');
				}
			} else {
				res.status(404).send('{"reason": "Invalid path"}');
				return;
			}
		} else {
			res.status(400).send('Unsupported HTTP method used');
		}
	});
}

function startServer(app, config) {
	var server = app.listen(config.SYSTEM.port, function() {
		var host = server.address().address;
		var port = server.address().port;

		logger.info("droxitApi server is listening at http://%s:%s", host, port);
	});
}

function gatherPlugins(conf, logger) {
	var plugins = {};
	for(name in conf) {
		var args = conf[name];
		if(!('params' in args))
			args.params = {};
		args.params.logger = logger;
		if('path' in args) {
			plugins[name] = {};
			require(args['path'])(plugins[name]);
			plugins[name].init(args['params']);
		} else {
			console.error('Path parameter missing from plugin definition');
			process.exit(1);
		}
	}

	return plugins;
}

// loogerName is only relevant if you want to start multiple servers from within the same
// process and want to separate their logging (which might be a good idea since their output would
// be indistinguishable otherwise)
exp.new = function(config, loggerName='droxit-api-server-logger') {
	var app = express();
	app.use(bodyParser.json());
	app.use(bodyParser.urlencoded({extended: true}));

	loglevel = "warn";
	if ('loglevel' in config.SYSTEM) {
		loglevel = config.SYSTEM.loglevel;
	}
	logOpts = {
		name: loggerName
	};

	if ('logfile' in config.SYSTEM) {
		logOpts.streams = [{
			path: config.SYSTEM.logfile,
			level: loglevel
		}];
	} else {
		logOpts.level = loglevel;
	}

	logger = bunyan.createLogger(logOpts);
	if (!('port' in config.SYSTEM)) {
		logger.fatal("missing port parameter in config file");
		process.exit(1);
	}

    if('PLUGINS' in config)
	    _plugins = gatherPlugins(config.PLUGINS, logger);

	setEndpoints(app, config, logger);
	startServer(app, config);

	return app;
};



src/startServer.js
/*
 * Helper script to start an API server with a given config file
 *
 * |------------------- OPEN SOURCE LICENSE DISCLAIMER -------------------|
 * |                                                                      |
 * | Copyright (C) 2019  droxIT GmbH - devs@droxit.de                     |
 * |                                                                      |
 * | This file is part of ROXconnector.                                   |
 * |                                                                      |
 * | ROXconnector is free software: you can redistribute it and/or modify |
 * | it under the terms of the GNU General Public License as published by |
 * | the Free Software Foundation, either version 3 of the License, or    |
 * | (at your option) any later version.                                  |
 * |                                                                      |
 * | This program is distributed in the hope that it will be useful,      |
 * | but WITHOUT ANY WARRANTY; without even the implied warranty of       |
 * | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the         |
 * | GNU General Public License for more details.                         |
 * |                                                                      |
 * | You have received a copy of the GNU General Public License           |
 * | along with this program. See also <http://www.gnu.org/licenses/>.    |
 * |                                                                      |
 * |----------------------------------------------------------------------|
 *
 */

var fs = require('fs');
var apiSrv = require('./roxconnector.js');

if (process.argv[2] == undefined) {
	console.error('The server needs a valid config file, passed as the first command line parameter, so please start the server again with a config file.');
	process.exit(1);
}

// read the config file, so the server only needs to execute the commands of the config file
fs.readFile(process.argv[2], 'utf-8', function(err, config) {
	if (!err) {
		config = JSON.parse(config);
		var server = apiSrv.new(config);
	} else {
		console.error(err);
		process.exit(1);
	}
});


src/roxconnector.js
/*
 * ROXconnector API server package
 * This server exposes a freely configurable REST service for interaction with the core system
 *
 * |------------------- OPEN SOURCE LICENSE DISCLAIMER -------------------|
 * |                                                                      |
 * | Copyright (C) 2019  droxIT GmbH - devs@droxit.de                     |
 * |                                                                      |
 * | This file is part of ROXconnector.                                   |
 * |                                                                      |
 * | ROXconnector is free software: you can redistribute it and/or modify |
 * | it under the terms of the GNU General Public License as published by |
 * | the Free Software Foundation, either version 3 of the License, or    |
 * | (at your option) any later version.                                  |
 * |                                                                      |
 * | This program is distributed in the hope that it will be useful,      |
 * | but WITHOUT ANY WARRANTY; without even the implied warranty of       |
 * | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the         |
 * | GNU General Public License for more details.                         |
 * |                                                                      |
 * | You have received a copy of the GNU General Public License           |
 * | along with this program. See also <http://www.gnu.org/licenses/>.    |
 * |                                                                      |
 * |----------------------------------------------------------------------|
 */

var fs = require('fs');
var express = require('express');
var http = require('http');
var bodyParser = require('body-parser');
var url = require('url');
var spawn = require('child_process').spawn;
var bunyan = require('bunyan');

var exp = module.exports;
var _plugins = {};

function setEndpoints(app, config, logger) {
	// handle errors gracefully
	app.use(function(err, req, res, next) {
		if (err) {
			logger.error(err);
			res.status(err.status).send(err.message);
		} else {
			next();
		}
	});
	// main http handling funcionality - available endpoints are defined in the config file
	app.use('/', function(req, res, next) {
		// setting headers to enable CORS (Cross Origin Ressource Sharing)
		res.header("Access-Control-Allow-Origin", req.get('Origin') || '*');
		res.header("Access-Control-Allow-Credentials", "true");
		res.header("Access-Control-Allow-Methods", "GET, HEAD, PUT, PATCH, POST, DELETE, OPTION");
		res.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Accept, Content-Type");

		res.append('Content-Type', 'application/json');
		logger.debug({
			request: req
		}, 'received request');

		if (req.method in config.REST) {
			// we only accept JSON encoded data bodies
			if (req.method === 'POST' || req.method === 'PUT') {
				if (!('content-type' in req.headers) ||
					(req.headers['content-type'] !== 'application/json' &&
						req.headers['content-type'] !== 'application/x-www-form-urlencoded')) {
					res.status(415).send('Invalid media type. MIME Type application/json required!');
					return;
				}
			}

			// check the config if the requested path is a valid endpoint
			var endpoint = config.REST[req.method];
			// get raw url
			var raw_url = url.parse(req.url).pathname;
			// remove trailing slash (if necessary)
			if (raw_url.charAt(raw_url.length - 1) == "/") {
				raw_url = raw_url.slice(0, -1);
			}
			var path = raw_url.slice(1).split('/');
			for (var p in path) {
				if (path[p] in endpoint)
					endpoint = endpoint[path[p]];
				else {
					res.status(404).send('{"reason": "Invalid path"}');
					return;
				}
			}

			var body = false;
			if (req.body && Object.keys(req.body).length !== 0) {
				body = req.body;
			}

			if ('handler' in endpoint) {
				if (endpoint.handler.type == 'process') {
					var jobConfig = endpoint.handler.command;
					var childProcess = spawn(jobConfig[0], jobConfig.slice(1));
					var childPid = childProcess.pid;
					var response = '';
					var error = '';

					if (body) {
						childProcess.stdin.write(JSON.stringify(body));
						childProcess.stdin.end();
					}

					childProcess.stdout.on('data', (data) => {
						response += data;
					});

					// logging stderr, to know if something bad happened
					childProcess.stderr.on('data', (data) => {
						error += data;
					});

					childProcess.on('exit', (code) => {
						logger.info({
							pid: childPid,
							exitCode: code
						}, 'child process exited');
						if (error) {
							try {
								logger.error({
									childPid: childPid,
									err_msg: error,
									requestBody: req.body,
									endpoint: '/' + path.join("/")
								}, 'handler process reported an error');
								err_obj = JSON.parse(error);
								res.status(err_obj.code).send(err_obj.reason);
							} catch (e) {
								res.status(500).send('{"reason": "internal server error"}');
							}
						} else
							res.send(response);
					});
				} else if (endpoint.handler.type == 'http') {
					if ('options' in endpoint.handler) {
						var data = '';
						// send a request to the handler
						var handlerReq = http.request(endpoint.handler.options, (r) => {
							var data = '';
							r.on('data', (chunk) => {
								data += chunk;
								logger.debug('received chunk: %s', chunk);
							});
							r.on('end', () => {
								logger.debug('sending data');
								res.send(data);
							});
						});
						handlerReq.on('error', (e) => {
							logger.error(e);
							res.status(500).send('{"reason": "internal server error"}');
						});
						if ('data' in endpoint.handler) {
							// use user data if any
							if (endpoint.handler.data === '-') {
								data = req.body;
							} else {
								data = endpoint.handler.data;
							}
						}

						if (data) {
							logger.debug({
								data: data
							}, "sending data to http handler");
							handlerReq.write(JSON.stringify(data));
						}
						handlerReq.end();
					} else {
						logger.error({
							path: "/" + path.join("/")
						}, "faulty configuration: endpoint is missing options field");
						res.status(500).send('{"reason": "internal server error"}');
					}
				} else if (endpoint.handler.type === 'plugin') {
					if (('name' in endpoint.handler) && ('function' in endpoint.handler)) {
						var pname = endpoint.handler.name;
						var func = endpoint.handler['function'];

						if ((pname in _plugins) && (func in _plugins[pname])) {
							var data = {};
							if (body) {
								data = body;
							}
							_plugins[pname][func](data, function(e, r) {
								if (e) {
									/*
									logger.error({
										plugin: pname,
										func: func,
										error: e
									}, "plugin returned an error"); */
									res.status(e.code).send(e.message);
								} else {
									res.send(r);
								}
							});
						} else {
							logger.error({
								path: "/" + path.join("/")
							}, "faulty configuration: plugin or function missing");
							res.status(500).send('{"reason": "internal server error"}');
						}
					} else {
						//TODO continue here
						logger.error({
							path: "/" + path.join("/")
						}, "faulty configuration: plugin handler is missing name or function argument");
						res.status(500).send('{"reason": "internal server error"}');
					}
					var func = endpoint.handler.function;
				} else {
					logger.error({
						path: "/" + path.join("/")
					}, "faulty configuration: invalid endpoint handler type");
					res.status(500).send('{"reason": "internal server error"}');
				}
			} else {
				res.status(404).send('{"reason": "Invalid path"}');
				return;
			}
		} else {
			res.status(400).send('Unsupported HTTP method used');
		}
	});
}

function startServer(app, config) {
	try {
		var server = app.listen(config.SYSTEM.port, function() {
			var host = server.address().address;
			var port = server.address().port;

			logger.info("droxitApi server is listening at http://%s:%s", host, port);
		}).on('error', (err) => {
			logger.fatal("could not start server - %s", err)
			process.exit(1)
		});
	} catch (e) {
		logger.fatal("could not start server - %s", e)
		process.exit(1)
	}
}

function gatherPlugins(conf, logger) {
	var plugins = {};
	for (name in conf) {
		var args = conf[name];
		if (!('params' in args))
			args.params = {};
		args.params.logger = logger;
		if ('path' in args) {
			plugins[name] = {};
			require(args['path'])(plugins[name]);
			plugins[name].init(args['params']);
		} else {
			console.error('Path parameter missing from plugin definition');
			process.exit(1);
		}
	}

	return plugins;
}

// loogerName is only relevant if you want to start multiple servers from within the same
// process and want to separate their logging (which might be a good idea since their output would
// be indistinguishable otherwise)
exp.new = function(config, loggerName = 'droxit-api-server-logger') {
	var app = express();
	if ('upload_limit' in config.SYSTEM) {
		var limit = {
			'limit': config.SYSTEM.upload_limit,
		}
		app.use(bodyParser.json(limit));
	} else {
		app.use(bodyParser.json());
	}
	app.use(bodyParser.urlencoded({
		extended: true
	}));

	loglevel = "warn";
	if ('loglevel' in config.SYSTEM) {
		loglevel = config.SYSTEM.loglevel;
	}
	logOpts = {
		name: loggerName
	};

	if ('logfile' in config.SYSTEM) {
		logOpts.streams = [{
			path: config.SYSTEM.logfile,
			level: loglevel
		}];
	} else {
		logOpts.level = loglevel;
	}

	logger = bunyan.createLogger(logOpts);
	if (!('port' in config.SYSTEM)) {
		logger.fatal("missing port parameter in config file");
		process.exit(1);
	}

	if ('PLUGINS' in config)
		_plugins = gatherPlugins(config.PLUGINS, logger);

	setEndpoints(app, config, logger);
	startServer(app, config);

	return app;
};
